### PR TITLE
Clean up errors on transaction execution

### PR DESCRIPTION
**Pull request by Kourin1996**
# Description

This PR cleans up errors around transaction execution including evm
- [x] Changes `fmt.Errorf` call to `errors.New` in error definitions
- [x] Renames `errReadOnly` to `errWriteProtection`
- [x] Adds `ErrInsufficientBalance` for error of insufficient balance on transfer
- [x] Adds `errGasUintOverflow` for overflow error on gas calculation
- [x] Add tests of error handling for the errors to need to be handled.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [x] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

**Comment by Kourin1996** [comment]
I changed the status to `draft` because I need to add some tests

**Review by zivkovicmilos — APPROVED** [review]
Clean and efficient 💯

**Review by lazartravica — APPROVED** [review]
The tests you added are 🔥! 

I'm okay with reusing errors between evm and runtime for the sake of better searching through different cases where a certain error might occur.

File: state/runtime/precompiled/precompiled.go [hunk 1]
```diff
@@ -96,11 +96,11 @@ func (p *Precompiled) Run(c *runtime.Contract, _ runtime.Host, config *chain.For
 	contract := p.contracts[c.CodeAddress]
 	gasCost := contract.gas(c.Input, config)
 
-	// In the case of not enough gas for precompiled execution we return ErrGasOverflow as opposed to ErrGasConsumed
+	// In the case of not enough gas for precompiled execution we return ErrGasOverflow as opposed to ErrOutOfGas
```
Comment (Vuksan) [inline state/runtime/precompiled/precompiled.go]: Can you modify this comment?
We don't return `ErrGasOverflow` as far as I can tell, we don't have that error in the codebase.

File: state/executor.go [hunk 2]
```diff
@@ -519,7 +523,7 @@ func (t *Transition) applyCreate(c *runtime.Contract, host runtime.Host) *runtim
 	if err := t.transfer(c.Caller, c.Address, c.Value); err != nil {
 		return &runtime.ExecutionResult{
 			GasLeft: gasLimit,
-			Err:     runtime.ErrNotEnoughFunds,
+			Err:     runtime.ErrInsufficientBalance,
```
Comment (Vuksan) [inline state/executor.go]: Why not just this?
As far as I see `transfer` method will return either this one `runtime.ErrInsufficientBalance` or some other error, so why not return correct error?
```suggestion
			Err:     err,
```

File: state/transition_test.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+			amount := big.NewInt(tt.amount)
+			err := transition.transfer(tt.from, tt.to, amount)
+
+			assert.Equal(t, tt.expectedErr, err)
+			if err == nil {
+				// should move balance
+				oldBalanceForFrom := big.NewInt(int64(tt.preState[tt.from].Balance))
+				oldBalanceForTo := big.NewInt(int64(tt.preState[tt.to].Balance))
+				newBalanceForFrom := new(big.Int).Sub(oldBalanceForFrom, amount)
+				newBalanceForTo := new(big.Int).Add(oldBalanceForTo, amount)
```
Comment (Vuksan) [inline state/transition_test.go]: Did you mean to fetch balances here?
The way it is now, test is calculating everything by itself, so only bigInt's `Sub` and `Add` are getting tested 🙂 
```suggestion
				newBalanceForFrom := transition.GetBalance(tt.from)
				newBalanceForTo := transition.GetBalance(tt.to)
```

File: state/runtime/evm/instructions_test.go [hunk 4]
```diff
@@ -90,3 +94,285 @@ func TestMStore(t *testing.T) {
...
 	assert.Len(t, s.memory, 1024+32)
 }
+
+type mockHostForCreate struct {
+	mockHost
+	nonce       uint64
+	callxResult *runtime.ExecutionResult
+}
+
+var _ runtime.Host = (*mockHostForCreate)(nil)
```
Comment (Vuksan) [inline state/runtime/evm/instructions_test.go]: What do you achieve with this line?
I don't understand 😕

File: state/runtime/evm/instructions_test.go [hunk 5]
```diff
@@ -90,3 +94,285 @@ func TestMStore(t *testing.T) {
...
+				memory: []byte{
+					byte(REVERT),
+				},
+				stop: true,
+				err:  errWriteProtection,
+			},
+			mockHost: &mockHostForCreate{},
+		},
+		{
+			name:     "should throw errOpCodeNotFound when op is CREATE2 and config.Constantinople is not disabled",
```
Comment (Vuksan) [inline state/runtime/evm/instructions_test.go]: I think you have a typo here:
```suggestion
			name:     "should throw errOpCodeNotFound when op is CREATE2 and config.Constantinople is disabled",
```

**Review by Vuksan — CHANGES_REQUESTED** [review]
Good job on the PR, especially instruction tests, they are excellent! 🙇 
Please take a look at the comments, no major changes are needed.

File: state/runtime/evm/instructions_test.go (hunk 5 above)
Comment (Kourin1996) [reply state/runtime/evm/instructions_test.go]: Fixed

File: state/runtime/evm/instructions_test.go (hunk 4 above)
Comment (Kourin1996) [reply state/runtime/evm/instructions_test.go]: Removed it

File: state/transition_test.go (hunk 3 above)
Comment (Kourin1996) [reply state/transition_test.go]: Fixed

File: state/executor.go (hunk 2 above)
Comment (Kourin1996) [reply state/executor.go]: Now it just returns errr

File: state/runtime/precompiled/precompiled.go (hunk 1 above)
Comment (Kourin1996) [reply state/runtime/precompiled/precompiled.go]: Fixed

**Review by Vuksan — APPROVED** [review]

**Pull request merged**